powertools-parameters/src/main/java/software/amazon/lambda/powertools/parameters/ParamManager.java

```diff
@@ -66,12 +97,13 @@ public static SSMProvider getSsmProvider() {
      * @return a {@link SecretsProvider}
      */
     public static SecretsProvider getSecretsProvider(SecretsManagerClient client) {
-        if (secretsProvider == null) {
+        if (!providers.containsKey(SecretsProvider.class)) {
             secretsProvider = SecretsProvider.builder()
                                     .withClient(client)
                                     .withCacheManager(cacheManager)
                                     .withTransformationManager(transformationManager)
                                     .build();
+            providers.put(SecretsProvider.class, secretsProvider);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **100 and 106**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **108**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.